Crafting menu tells you which colour of crayon is needed

cacogen opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

I named the crayons.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

The crafting menu will tell you which colour of crayon is needed in a given recipe now. Helpful when collecting ingredients for the crazy burger.

Fixes #56906
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl: cacogen
qol: Crafting menu will now tell you which colour of crayon is needed (useful for recipes like the crazy burger)
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

cacogen committed:
Gives crayons names

MrDoomBringer commented on code/game/objects/items/crayons.dm:
```diff
@@ -502,62 +502,71 @@
 			return G
 
 /obj/item/toy/crayon/red
```
Would it be feasible to set the name on Initialize using the `crayon_color` variable?
You'd have to add case checking for things like the mime color, but I imagine it would be cleaner

cacogen replied:
It does that already but the crafting menu only cares about the initial name.

MrDoomBringer replied:
oh i see lol

Mothblocks commented:
Put the "fixes" in your description so it will automatically be closed.

Cyberboss reviewed (changes requested):
You mentioned that crayons set their names based on crayon_color but this isn't valid because initial is used. Could you, instead of having Initialize set the name, validate that it is correct for crayon_color. Have it stack_trace if it isn't. That way, if someone adds a new color, we have validation that they remembered to set an explicitly correct name.

Mothblocks commented:
> Could you, instead of having Initialize set the name, validate that it is correct for crayon_color. Have it stack_trace if it isn't. That way, if someone adds a new color, we have validation that they remembered to set an explicitly correct name.

A unit test sounds smarter for this instead.

cacogen commented:
How do I do a unit test?

MrDoomBringer commented:
https://github.com/tgstation/tgstation/blob/master/code/modules/unit_tests/README.md

cacogen committed:
Adds a unit test and test for said unit test

Mothblocks commented on code/modules/unit_tests/crayons.dm:
```diff
@@ -0,0 +1,12 @@
+			bad_crayons_list += "[this_crayon.type]"
+		if(!L.len)
+			return //passed!
+		Fail("The following crayons do not have their crayon_color in their initial name: [L.Join(",")])
```
Don't do this. Just call `Fail` whenever you get to an invalid name. It will continue the test.

Mothblocks commented on code/modules/unit_tests/crayons.dm:
```diff
@@ -0,0 +1,12 @@
+
+/datum/unit_test/crayon_naming/Run()
+	var/list/bad_crayons_list = list()
+	for(var/i in typesof(/obj/item/toy/crayon))
+		var/obj/item/toy/crayon/this_crayon = i
```
```suggestion
	for(var/obj/item/toy/crayon/this_crayon as anything in typesof(/obj/item/toy/crayon))
```

Mothblocks commented on code/modules/unit_tests/crayons.dm:
```diff
@@ -0,0 +1,8 @@
+		if(is_type(this_crayon, /obj/item/toy/crayon/spraycan))
+			continue
+		if(!findtext("[this_crayon]","[this_crayon.crayon_color]"))
+			Fail("[this_crayon.type] does not have its crayon_color ([crayon_color]) in its initial name ([this_crayon]).")
```
`in its initial name ([this_crayon])."`

`this_crayon` is a typepath. This will display `/obj/item/toy/crayon/red`, for instance.

Mothblocks commented:
Please read the README @MrDoomBringer posted on how to run these tests locally.

cacogen committed:
Why did Visual Studio Code compile this successfully

cacogen commented:
@Mothblocks 

I don't know why God won't let me do this
```
[22:07:10] Runtime in crayons.dm,7: Cannot read /obj/item/toy/crayon/spraycan/... (/obj/item/toy/crayon/spraycan/borg).crayon_color
  proc name: Run (/datum/unit_test/crayon_naming/Run)
  src: /datum/unit_test/crayon_naming (/datum/unit_test/crayon_naming)
  call stack:
  /datum/unit_test/crayon_naming (/datum/unit_test/crayon_naming): Run()
  RunUnitTests()
  /datum/callback (/datum/callback): InvokeAsync()
```

```dm
/datum/unit_test/crayon_naming/Run()
	for(var/obj/item/toy/crayon/this_crayon as anything in typesof(/obj/item/toy/crayon))
		if(istype(this_crayon, /obj/item/toy/crayon/spraycan))
			continue
		if(!findtext(initial(this_crayon.name),this_crayon.crayon_color))
			Fail("[this_crayon] does not have its crayon_color ([this_crayon.crayon_color]) in its initial name ([initial(this_crayon.name)]).")
```

Mothblocks commented:
Change `istype` to `ispath`--`istype` only works on instantiated datums.

Mothblocks commented on code/modules/unit_tests/crayons.dm:
```diff
@@ -0,0 +1,8 @@
+///Makes sure that crayons have their crayon_color in their initial name.
```
```suggestion
/// Makes sure that crayons have their crayon_color in their initial name.
/datum/unit_test/crayon_naming
```

That way, it shows up in the documentation.

cacogen committed:
We did it

cacogen committed:
Really went back to kindergarten on this one

- Removes the use of Initialize to set a crayon's proper name given the unit test disallows this
- Gives mime crayon its name back and gives the base crayon the proper name for its colour

Mothblocks commented on code/modules/unit_tests/crayons.dm:
```diff
@@ -0,0 +1,10 @@
+/// Makes sure that crayons have their crayon_color in their initial name (to differentiate them in the crafting menu).
```
That suggestion wasn't just to put a space before the comment, it was so that it has something to document:

```suggestion
/// Makes sure that crayons have their crayon_color in their initial name (to differentiate them in the crafting menu).
/datum/unit_test/crayon_naming
```

Otherwise, you're documenting `Run()`.

Other than that, the test seems fine.

cacogen committed:
Documents the unit test and not Run()

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

Mothblocks commented:
Mark it as undrafted when you're done.

Cyberboss commented:
Unit tests are not a player facing change and should not be in the changelog

Mothblocks commented:
Thanks for the tests! I removed the mention of the test from the changelog.

cacogen commented:
Code changes aren't player-facing by definition and I added that because this simple fix PR took fucking forever because of the unit test one of you made me do and I felt I deserved the points for it

Mothblocks commented:
The points are not a result of the changelog, they're of your labels.

I gave you the Unit Tests label, which I believe gives 6 points.

cacogen commented:
I have to stop assuming things

MrDoomBringer commented:
bruh
